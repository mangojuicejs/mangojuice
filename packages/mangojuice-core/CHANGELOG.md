# Change Log

All notable changes to this project will be documented in this file.
See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="0.1.3"></a>
## 0.1.3 (2017-07-12)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))
* **core:** recursive nesting works now ([c462be5](https://github.com/mangojuicejs/mangojuice/commit/c462be5))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.3"></a>
## 0.1.3 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)



<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2"></a>
## 0.1.2 (2017-07-08)


### Bug Fixes

* **core:** fixed subscriptions handling ([63915a4](https://github.com/mangojuicejs/mangojuice/commit/63915a4))
* **core:** move back cmd obj id after handling to make it reusable ([3b6159c](https://github.com/mangojuicejs/mangojuice/commit/3b6159c))


### Features

* **core:** added computed fields memoization ([d2ae2d7](https://github.com/mangojuicejs/mangojuice/commit/d2ae2d7))
* **core:** computed fields, initial impl without lazy computing ([c5b303a](https://github.com/mangojuicejs/mangojuice/commit/c5b303a))
* **core:** made shared block optional ([cfee213](https://github.com/mangojuicejs/mangojuice/commit/cfee213))



<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.1"></a>
## 0.1.2-beta.1 (2017-07-03)



<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)




<a name="0.1.2-beta.0"></a>
## 0.1.2-beta.0 (2017-07-03)



<a name="0.1.1"></a>
## 0.1.1 (2017-07-03)
