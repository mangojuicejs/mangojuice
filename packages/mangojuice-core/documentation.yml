toc:
  - name: Basic
    description: |
      The most common APIs that you will work with most of the time
    children:
      - Block
      - LogicBase
      - cmd
      - run
      - mount
      - task
      - cancel
      - child
      - depends
  - name: Extras
    description: |
      Other pieces of MangoJuice
    children:
      - Process
      - Command
      - TaskMeta
      - DefaultLogger
      - ComputedField
      - Mounter
      - defineCommand
      - ensureCommand
      - decorateLogic
      - handleBefore
      - handleAfter
      - handle
      - call
      - logicOf
      - observe
      - procOf
      - bind
      - hydrate
      - delay
