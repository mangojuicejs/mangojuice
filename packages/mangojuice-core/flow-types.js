// @flow
type CommandType<M> = {
  is: (...args: any) => bool,
  model: M,
};

type TaskObject = {
  success: (cmd: any) => TaskObject,
  fail: (cmd: any) => TaskObject,
  args: (...args: Array<any>) => TaskObject,
  engine: (exec: any) => TaskObject
};

type DependsType = {
  compute: (Function) => DependsType
};

type LogicConfig<MetaType> = {
  initCommands?: any,
  meta?: MetaType
};

type ConfiguredLogic = {
};

type BlockType = {
  createModel: (...args: any) => Object,
  Logic: Class<*>,
  View?: any
};

declare module "mangojuice-core" {
  declare export function cmd(): any;

  declare export function debounce(): any;

  declare export function throttle(): any;

  declare export function task(fn: Function): TaskObject;

  declare export function delay(ms: number): Promise<void>;

  declare export function cancel(cmd: any): any;

  declare export function logicOf<T>(model: Object, clazz?: Class<T>): T;

  declare export function child<T>(logicClass: Class<{ +config?: (...args: T) => any }>, ...args: T): ConfiguredLogic;

  declare export function depends(...args: any): DependsType;

  declare export function observe(model: Object, destroyeded: Promise<any>, handler: Function): void;

  declare export function run(block: BlockType, opts?: Object): any;

  declare export function bind(block: BlockType, opts?: Object): any;

  declare export function hydrate(block: BlockType, model: Object): any;

  declare export var utils: any;

  declare export class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
    +model: ModelType;
    +shared: SharedType;
    +meta: MetaType;

    hub(cmd: CommandType<*>): ?Promise<any>;

    hubBefore(cmd: CommandType<*>): ?Promise<any>;

    hubAfter(cmd: CommandType<*>): ?Promise<any>;

    port(exec: (cmd: any) => Promise<any>, destroyed: Promise<any>): ?Promise<any>;

    children(): {[k: $Enum<ModelType>]: Class<*> | ConfiguredLogic};

    computed(): {[k: $Enum<ModelType>]: (() => any) | DependsType};

    config(...args: any): LogicConfig<MetaType>;
  }

  declare export default {|
    +LogicBase: typeof LogicBase,
    +cmd: typeof cmd,
    +debounce: typeof debounce,
    +throttle: typeof throttle,
    +task: typeof task,
    +delay: typeof delay,
    +cancel: typeof cancel,
    +logicOf: typeof logicOf,
    +child: typeof child,
    +depends: typeof depends,
    +observe: typeof observe,
    +run: typeof run,
    +bind: typeof bind,
    +hydrate: typeof hydrate,
    +utils: typeof utils
  |};
}
