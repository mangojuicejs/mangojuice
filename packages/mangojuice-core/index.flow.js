// @flow

type NestType<A, T, Y, U> = {
  +logic: A,
  args: (c1?: T, c2?: Y, c3?: U) => NestType<A, T, Y, U>,
  handler: () => NestType<A, T, Y, U>,
  singleton: (?Boolean) => NestType<A, T, Y, U>
};

type DependsType = {
  compute: (Function) => DependsType
};

interface ConfigFunction<MetaType, T, Y, U> {
  +config?: (c1?: T, c2?: Y, c3?: U) => ConfigType<MetaType>;
};

type ConfigType<M> = {
  initCommands?: any,
  meta?: M
};

declare class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
  +model: ModelType;
  +shared: SharedType;
  +meta: MetaType;
  +destroy: Promise<any>;

  port(): ?Promise<any>;

  children(): {[key: $Enum<ModelType>]: NestType<*, *, *, *>};

  computed(): {[k: $Enum<ModelType>]: () => any | DependsType};

  exec(): Promise<any>;

  depends(...deps: Array<Object>): DependsType;

  nest<T, Y, U, L: ConfigFunction<MetaType, T, Y, U>&LogicBase<*,SharedType,*>>(logic: L): NestType<L, T, Y, U>;
}
