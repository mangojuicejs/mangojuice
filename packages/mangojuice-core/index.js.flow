// @flow

// Cmd
//
declare export var Cmd: {
  update: any,
  batch: any,
  execLatest: any,
  execEvery: any,
  debounce: any,
  throttle: any
};

// Task
//
type TaskObject = {
  success: (cmd: any) => TaskObject,
  fail: (cmd: any) => TaskObject
};
declare export var Task: {
  create: (Function) => TaskObject;
};


// Run
//
declare export var Run: any;


// LogicBase
//
type NestType<A, T, Y, U> = {
  +logic: A,
  args: (c1?: T, c2?: Y, c3?: U) => NestType<A, T, Y, U>,
  handler: () => NestType<A, T, Y, U>,
  singleton: (?Boolean) => NestType<A, T, Y, U>
};

type DependsType = {
  compute: (Function) => DependsType
};

interface ConfigFunction<MetaType, T, Y, U> {
  +config?: (c1?: T, c2?: Y, c3?: U) => ConfigType<MetaType>;
};

type ConfigType<M> = {
  initCommands?: any,
  meta?: M
};

declare export class LogicBase<ModelType = any, SharedType = any, MetaType = any> {
  +model: ModelType;
  +shared: SharedType;
  +meta: MetaType;
  +destroy: Promise<any>;

  port(): ?Promise<any>;

  children(): {[key: $Enum<ModelType>]: NestType<*, *, *, *>};

  computed(): {[k: $Enum<ModelType>]: () => any | DependsType};

  exec(): Promise<any>;

  depends(...deps: Array<Object>): DependsType;

  nest<T, Y, U, L: ConfigFunction<MetaType, T, Y, U>&LogicBase<*,SharedType,*>>(logic: L): NestType<L, T, Y, U>;
}
