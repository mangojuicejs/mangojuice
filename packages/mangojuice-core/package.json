{
  "name": "mangojuice-core",
  "version": "0.1.6",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs"
  }
}
