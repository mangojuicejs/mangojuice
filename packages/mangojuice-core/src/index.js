import Process from './classes/Process';
import Command from './classes/Command';
import LogicBase from './classes/LogicBase';
import DefaultLogger from './classes/DefaultLogger';
import cmd from './core/cmd/cmd';
import cancel from './core/cmd/cancel';
import defineCommand from './core/cmd/defineCommand';
import ensureCommand from './core/cmd/ensureCommand';
import child from './core/logic/child';
import depends from './core/logic/depends';
import observe from './core/logic/observe';
import { handle, handleAfter, handleBefore } from './core/logic/handle';
import logicOf from './core/logic/logicOf';
import procOf from './core/logic/procOf';
import decorateLogic from './core/logic/decorateLogic';
import delay from './core/task/delay';
import task from './core/task/task';
import callTask from './core/task/callTask';
import { CANCEL } from './classes/AsyncTask';
import run from './core/run/run';
import bind from './core/run/bind';
import mount from './core/run/mount';
import hydrate from './core/run/hydrate';
import * as utils from './core/utils';
import * as config from './config';


export {
  // Constants
  CANCEL,
  // Classes
  Process,
  Command,
  LogicBase,
  DefaultLogger,
  // Core
  cmd,
  cancel,
  defineCommand,
  ensureCommand,
  child,
  depends,
  observe,
  handle,
  handleAfter,
  handleBefore,
  logicOf,
  procOf,
  decorateLogic,
  delay,
  task,
  callTask,
  run,
  bind,
  mount,
  hydrate,
  utils,
  config
};
