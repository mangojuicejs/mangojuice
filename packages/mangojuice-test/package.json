{
  "name": "mangojuice-test",
  "version": "2.0.0-y.0",
  "description": "",
  "author": {
    "name": "Artem Artemev",
    "email": "viiv.c58@gmail.com"
  },
  "bugs": {
    "url": "https://github.com/mangojuicejs/mangojuice/issues"
  },
  "homepage": "https://github.com/mangojuicejs/mangojuice#readme",
  "repository": "https://github.com/mangojuicejs/mangojuice",
  "license": "MIT",
  "main": "index.js",
  "rollup": {
    "moduleName": "Mjs.Test",
    "bundledDependencies": [],
    "moduleGlobals": {
      "mangojuice": "Mjs"
    }
  },
  "dependencies": {
    "mangojuice-core": "2.0.0-y.0"
  }
}
