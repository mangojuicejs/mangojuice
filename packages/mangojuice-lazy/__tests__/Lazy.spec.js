import { cmd, logicOf, child, delay, handleLogicOf } from "mangojuice-core";
import { runWithTracking } from "mangojuice-test";
import lazyUtils from "mangojuice-lazy";

process.on('unhandledRejection', error => {
  // Will print "unhandledRejection err is not defined"
  console.log(error);
});

const createMockBlockResolver = (Block) => {
  let resolve = null;
  const resolvePromise = new Promise(r => resolve = r);
  const resolver = (callback) => setTimeout(() => {
    callback(Block);
    resolve();
  }, 100);
  return { resolved: resolvePromise, resolver };
};

describe('Lazy block loading', () => {
  const AppBlock = {
    createModel: () => ({ e: 0, f: 4, g: 6, c: null }),
    Logic: class AppBlock {
      config() {
        return {
          initCommands: this.CmdFromInit,
          meta: 'test-meta'
        }
      }
      port(exec) {
        return exec(this.CmdFromPort);
      }
      computed() {
        return {
          e: () => this.model.f + this.model.g
        };
      }
      @cmd SetField(name, value) {
        return { [name]: value };
      }
      @cmd CmdFromPort() {}
      @cmd CmdFromInit() {
        return { c: this.meta }
      }
    }
  };

  it('shuold do nothing if view is not triggered', async () => {
    const resolver = jest.fn();
    const LazyBlock = lazyUtils.createLazyBlock(resolver);
    const { app, shared, commandNames, errors } = await runWithTracking({
      app: LazyBlock
    });

    expect(commandNames).toEqual([]);
    expect(resolver).not.toBeCalled();
  });

  it('should resolve block when view reuqested', async () => {
    const { resolver, resolved } = createMockBlockResolver(AppBlock);
    const LazyBlock = lazyUtils.createLazyBlock(resolver);
    const { app, shared, commandNames, errors } = await runWithTracking({
      app: LazyBlock
    });

    LazyBlock.View({ model: app.model });
    await resolved;
    await delay(10);

    expect(commandNames).toEqual([
      'AppBlock.CmdFromPort',
      'AppBlock.CmdFromInit'
    ]);
  });

  it('should resolve block when any command executed', async () => {
    const { resolver, resolved } = createMockBlockResolver(AppBlock);
    const LazyBlock = lazyUtils.createLazyBlock(resolver);
    const { app, shared, commandNames, errors } = await runWithTracking({
      app: LazyBlock
    });

    await app.proc.exec(logicOf(app.model).SetField('f', 6));
    await resolved;

    expect(commandNames).toEqual([
      'LazyBlock.SetField.Lazy',
      'AppBlock.CmdFromPort',
      'AppBlock.CmdFromInit',
      'AppBlock.SetField'
    ]);
    expect(app.model).toEqual({ e: 12, f: 6, g: 6, c: 'test-meta' });
  });

  it('should works on re-render after resolve', async () => {
    const BlockChild = {
      createModel: (props) => ({ ...props }),
      Logic: class AppLogic {
        config() {
          return {
            initCommands: this.CmdFromInit,
            meta: 'test-meta'
          }
        }
        port(exec) {
          return exec(this.CmdFromPort);
        }
        @cmd CmdFromPort() {}
        @cmd CmdFromInit() {
          return { c: this.meta }
        }
        @cmd SetField(name, value) {
          return { [name]: value };
        }
      }
    };
    const { resolver, resolved } = createMockBlockResolver(BlockChild);
    const LazyBlock = lazyUtils.createLazyBlock(resolver);
    const BlockParent = {
      createModel: () => ({ child: null }),
      Logic: class BlockParent {
        children() {
          return { child: child(LazyBlock.Logic) };
        }
        @cmd SetField(name, value) {
          return { [name]: value };
        }
      }
    };
    const { app, shared, commandNames, errors } = await runWithTracking({
      app: BlockParent
    });

    await app.proc.exec(logicOf(app.model).SetField('child', LazyBlock.createModel({'b': 321})));
    await app.proc.exec(logicOf(app.model.child).SetField('a', 123));
    await resolved;
    await app.proc.exec(logicOf(app.model).SetField('child', LazyBlock.createModel()));

    expect(commandNames).toEqual([
      'BlockParent.SetField',
      'LazyBlock.SetField.Lazy',
      'AppLogic.CmdFromPort',
      'AppLogic.CmdFromInit',
      'AppLogic.SetField',
      'BlockParent.SetField',
      'AppLogic.CmdFromPort',
      'AppLogic.CmdFromInit',
    ]);
    expect(app.model.child).toEqual({ c: 'test-meta' });
  });

  it('should update model after resolve', async () => {
    const BlockChild = {
      createModel: (props) => ({ ...props }),
      Logic: class AppLogic {
        @cmd SetField(name, value) {
          return { [name]: value };
        }
      }
    };
    const { resolver, resolved } = createMockBlockResolver(BlockChild);
    const LazyBlock = lazyUtils.createLazyBlock(resolver);
    const BlockParent = {
      createModel: () => ({ child: null }),
      Logic: class BlockParent {
        children() {
          return { child: child(LazyBlock.Logic) };
        }
        @cmd SetField(name, value) {
          return { [name]: value };
        }
      }
    };
    const { app, shared, commandNames, errors } = await runWithTracking({
      app: BlockParent
    });

    const lazyModel = LazyBlock.createModel({'b': 321});
    expect(lazyModel).toEqual({ __args: [{'b': 321}] });

    await app.proc.exec(logicOf(app.model).SetField('child', lazyModel));
    await app.proc.exec(logicOf(app.model.child).SetField('a', 123));
    await resolved;
    expect(app.model.child).toEqual({ a: 123, b: 321 });

    const realModel = LazyBlock.createModel({'b': 321});
    expect(realModel).toEqual({'b': 321});

    await app.proc.exec(logicOf(app.model).SetField('child', LazyBlock.createModel()));
    expect(app.model.child).toEqual({});
  });

  it('should fill model with actual data even if the model was detached', async () => {
    const BlockChild = {
      createModel: (props) => ({ ...props }),
      Logic: class AppLogic {
        @cmd SetField(name, value) {
          return { [name]: value };
        }
      }
    };
    const { resolver, resolved } = createMockBlockResolver(BlockChild);
    const LazyBlock = lazyUtils.createLazyBlock(resolver);
    const BlockParent = {
      createModel: () => ({ child: null }),
      Logic: class BlockParent {
        children() {
          return { child: child(LazyBlock.Logic) };
        }
        @cmd SetField(name, value) {
          return { [name]: value };
        }
      }
    };
    const { app, shared, commandNames, errors } = await runWithTracking({
      app: BlockParent
    });

    const lazyModel = LazyBlock.createModel({'b': 321});
    app.proc.exec(logicOf(app.model).SetField('child', lazyModel));
    const changeLazy = app.proc.exec(logicOf(app.model.child).SetField('a', 123));
    app.proc.exec(logicOf(app.model).SetField('child', null));
    await resolved;
    await changeLazy;

    expect(lazyModel).toEqual({ b: 321 });
    expect(app.model.child).toEqual(null);
  });

  it('should make lazy command nonhandlable but original handled', async () => {
    const BlockChild = {
      createModel: (props) => ({ ...props }),
      Logic: class AppLogic {
        @cmd SetField(name, value) {
          return { [name]: value };
        }
      }
    };
    const { resolver, resolved } = createMockBlockResolver(BlockChild);
    const LazyBlock = lazyUtils.createLazyBlock(resolver);
    const handled = jest.fn();
    const BlockParent = {
      createModel: () => ({ child: null }),
      Logic: class BlockParent {
        children() {
          return { child: child(LazyBlock.Logic) };
        }
        hub(cmd) {
          if (cmd.is(logicOf(this.model.child).SetField)) {
            handled(cmd);
            return this.HandleChild;
          }
        }
        @cmd SetField(name, value) {
          return { [name]: value };
        }
        @cmd HandleChild(cmd) {
        }
      }
    };
    const { app, shared, commandNames, errors } = await runWithTracking({
      app: BlockParent
    });

    await app.proc.exec(logicOf(app.model).SetField('child', LazyBlock.createModel({'b': 321})));
    await app.proc.exec(logicOf(app.model.child).SetField('a', 123));
    await resolved;

    expect(commandNames).toEqual([
      'BlockParent.SetField',
      'LazyBlock.SetField.Lazy',
      'AppLogic.SetField',
      'BlockParent.HandleChild'
    ]);
    expect(handled).toHaveBeenCalledTimes(1);
    expect(handled.mock.calls[0][0].func).toBeDefined();
    expect(handled.mock.calls[0][0].func).toEqual(BlockChild.Logic.prototype.SetField.func);
  });

  it('should maintain shared block binsing after module resolving', async () => {
    // TODO
    // 1. Add shared block to the app, lazy with hub with
    // 2. Run lazy and shared
    // 3. Resolve lazy
    // 4. Run some command in shared
    // 5. Hub should be called only once (no shared double-binding after resolution)
  });
});
