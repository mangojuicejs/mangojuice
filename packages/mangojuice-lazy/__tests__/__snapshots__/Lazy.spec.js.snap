// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lazy block loading should NOT fill model with actual data even if the model was detached 1`] = `
Array [
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
]
`;

exports[`Lazy block loading should NOT fill model with actual data even if the model was detached 2`] = `
Object {
  "__lazy": true,
}
`;

exports[`Lazy block loading should resolve block when lazy message executed 1`] = `
Array [
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
  Object {
    "args": Array [],
    "c": null,
    "e": 0,
    "f": 4,
    "g": 6,
  },
]
`;

exports[`Lazy block loading should resolve block when lazy message executed 2`] = `
Object {
  "args": Array [],
  "c": null,
  "e": 0,
  "f": 4,
  "g": 6,
}
`;

exports[`Lazy block loading should resolve block when lazy message executed 3`] = `
Array [
  Array [
    Message {
      "args": Array [
        1,
        2,
        3,
      ],
    },
  ],
]
`;

exports[`Lazy block loading should resolve block when view reuqested 1`] = `
Array [
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
  Object {
    "args": Array [],
    "c": null,
    "e": 0,
    "f": 4,
    "g": 6,
  },
]
`;

exports[`Lazy block loading should resolve block when view reuqested 2`] = `
Object {
  "args": Array [],
  "c": null,
  "e": 0,
  "f": 4,
  "g": 6,
}
`;

exports[`Lazy block loading should resolve multiple lazy children 1`] = `
Array [
  Object {
    "anotherChild": ChildCmd {
      "createArgs": Array [
        "ccc",
      ],
      "logicClass": [Function],
    },
    "child": Array [
      ChildCmd {
        "createArgs": Array [
          "a",
          "b",
          "c",
        ],
        "logicClass": [Function],
      },
      ChildCmd {
        "createArgs": Array [
          "aaa",
        ],
        "logicClass": [Function],
      },
    ],
  },
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
  Object {
    "anotherChild": ChildCmd {
      "logicClass": [Function],
      "updateMsg": Message {
        "args": Array [
          1,
          2,
          3,
        ],
        "name": "TestMessage",
      },
    },
  },
  Object {
    "child": Array [
      ChildCmd {
        "logicClass": [Function],
        "updateMsg": Message {
          "args": Array [
            1,
            2,
            3,
          ],
          "name": "TestMessage",
        },
      },
      ChildCmd {
        "logicClass": [Function],
        "updateMsg": Message {
          "args": Array [
            133,
          ],
          "name": "TestMessage",
        },
      },
    ],
  },
  Object {
    "args": Array [
      "ccc",
    ],
    "c": null,
    "e": 0,
    "f": 4,
    "g": 6,
  },
  Object {
    "args": Array [
      "a",
      "b",
      "c",
    ],
    "c": null,
    "e": 0,
    "f": 4,
    "g": 6,
  },
  Object {
    "args": Array [
      "aaa",
    ],
    "c": null,
    "e": 0,
    "f": 4,
    "g": 6,
  },
]
`;

exports[`Lazy block loading should resolve multiple lazy children 2`] = `
Object {
  "anotherChild": Object {
    "args": Array [
      "ccc",
    ],
    "c": null,
    "e": 0,
    "f": 4,
    "g": 6,
  },
  "child": Array [
    Object {
      "args": Array [
        "a",
        "b",
        "c",
      ],
      "c": null,
      "e": 0,
      "f": 4,
      "g": 6,
    },
    Object {
      "args": Array [
        "aaa",
      ],
      "c": null,
      "e": 0,
      "f": 4,
      "g": 6,
    },
  ],
}
`;

exports[`Lazy block loading should works on re-render after resolve 1`] = `
Array [
  Object {
    "child": ChildCmd {
      "createArgs": Array [
        "a",
        "b",
        "c",
      ],
      "logicClass": [Function],
    },
  },
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
  Object {
    "child": ChildCmd {
      "logicClass": [Function],
      "updateMsg": Message {
        "args": Array [
          1,
          2,
          3,
        ],
        "name": "TestMessage",
      },
    },
  },
  Object {
    "args": Array [
      "a",
      "b",
      "c",
    ],
    "c": null,
    "e": 0,
    "f": 4,
    "g": 6,
  },
  Object {
    "child": ChildCmd {
      "logicClass": [Function],
      "updateMsg": Message {
        "args": Array [
          "111",
        ],
      },
    },
  },
]
`;

exports[`Lazy block loading should works on re-render after resolve 2`] = `
Object {
  "args": Array [
    "a",
    "b",
    "c",
  ],
  "c": null,
  "e": 0,
  "f": 4,
  "g": 6,
}
`;

exports[`Lazy block loading should works on re-render after resolve 3`] = `
Array [
  Array [
    Message {
      "args": Array [
        1,
        2,
        3,
      ],
    },
  ],
  Array [
    Message {
      "args": Array [
        "111",
      ],
    },
  ],
]
`;

exports[`Lazy block loading shuold do nothing if view is not triggered 1`] = `
Array [
  Array [
    undefined,
    Object {
      "__lazy": true,
    },
  ],
  Object {
    "__lazy": true,
  },
]
`;
